nova/virt/vmwareapi/driver.py

```diff
@@ -753,6 +760,290 @@ def _server_group_sync_loop(self, compute_host):
+            # new volume attachments
+            # This cannot be done in pre-check, as the new volume attachments
+            # are created prior the live-migration
+            volumes = self._get_checked_volumes(context, instance,
+                required_volume_attributes)
+            self._set_vif_infos(migrate_data, dest_session)
+            self._vmops.live_migration(instance, migrate_data, volumes)
+            LOG.info("Migration operation completed", instance=instance)
+            post_method(context, instance, dest, block_migration, migrate_data)
+        except Exception:
+            LOG.exception("Failed due to an exeption", instance=instance)
+            with excutils.save_and_reraise_exception():
+                # We are still in the task-state migrating, so cannot
+                # recover the DRS settings. We rely on the sync to do that
+                LOG.debug("Calling live migration recover_method "
+                          "for instance: %s", instance["name"],
+                          instance=instance)
+                recover_method(context, instance, dest, migrate_data)
+
+    def _get_volume_mappings(self, context, instance):
+        """Returns a mapping for all the devices with volumes to
+        their connection_info.data, where possible.
+        It is up to the caller to verify that the required information
+        is available.
+        It uses the _current_ block-device mapping.
+        """
+        volume_infos = []
+        for bdm in objects.BlockDeviceMappingList.get_by_instance_uuid(
+                    context, instance.uuid):
+            if not bdm.is_volume:
+                continue
+            # If we have volume_id, `map_volumes_to_devices` will map it to
+            # a device.
+            data = {"volume_id": bdm.volume_id}
+            try:
+                connection_info = jsonutils.loads(bdm.connection_info)
+                data.update(connection_info.get("data", {}))
+            except ValueError as e:
+                message = ("Could not parse connection_info for volume {}."
+                    " Reason: {}"
+                ).format(bdm.volume_id, e)
+                LOG.warning(message, instance=instance)
+
+            # Normalise the datastore reference
+            # As it depends on the caller, if actually need the
+            # value, we do not raise an error here
+            datastore = data.get("datastore") or \
+                            data.get("ds_ref_val")
+            if datastore:
+                data["datastore_ref"] = vim_util.get_moref(datastore,
+                                                        "Datastore")
+            volume_infos.append(data)
+        return self._volumeops.map_volumes_to_devices(instance, volume_infos)
+
+    def _set_vif_infos(self, migrate_data, session):
+        if 'vifs' not in migrate_data or not migrate_data.vifs:
+            migrate_data.vif_infos = []
+        else:
```

Comment:
You moved it to the top, but didn't add a `return` to get rid of the indentation and the `else`. That would have been my idea for moving it to the top of the function.

Reply:
I'll add the return